docs: add NG0403.md for Module Bootstrapped Without Bootstrap Component

robertIsaac opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
the error is missing from the errors list

Issue Number: [47985](https://github.com/angular/angular/issues/47985)


## What is the new behavior?
the error is added now

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

robertIsaac committed:
docs: add NG0403.md for Bootstrapped NgModule doesn't specify which component to initialize error

- update `errors.ts` to annotate the error NG0403, so that the runtime can add a link to that guide when an error is thrown
- update `application_ref_spec.ts` to include the new link of the error
- update `errors.md` as a result of running `yarn bazel test packages/core/test`

Fixes #47985

jessicajaniuk commented on aio/content/errors/NG0403.md:
````diff
@@ -0,0 +1,30 @@
+```
+
+@debugging
+Did you mistakenly removed the `bootstrap` property of your `AppModule` or remove the `AppComponent` from it?
````
```suggestion
Did you mistakenly remove the `bootstrap` property of your `AppModule` or remove the `AppComponent` from it?
```

robertIsaac replied:
[AndrewKushnir](https://github.com/AndrewKushnir)

jessicajaniuk reviewed (approved):
LGTM other than the one grammatical error. Thanks!

reviewed-for: public-api

AndrewKushnir commented on aio/content/errors/NG0403.md:
```diff
@@ -0,0 +1,30 @@
+@name Module Bootstrapped Without Bootstrap Component
```
```suggestion
@name Bootstrapped NgModule doesn't specify which component to initialize
```

robertIsaac replied:
done

AndrewKushnir commented on aio/content/errors/NG0403.md:
```diff
@@ -0,0 +1,30 @@
+@name Module Bootstrapped Without Bootstrap Component
+@category runtime
+@shortDescription Module AppModule Was Bootstrapped, But It Does Not Declare "@NgModule.bootstrap" Components Nor A "ngDoBootstrap" Method Please define one of these.
```
```suggestion
@shortDescription An NgModule that was used for bootstrapping does not specify which component should be initialized.
```

robertIsaac replied:
@AndrewKushnir changes will not use this sentence at all

AndrewKushnir commented on aio/content/errors/NG0403.md:
```diff
@@ -0,0 +1,30 @@
+@shortDescription Module AppModule Was Bootstrapped, But It Does Not Declare "@NgModule.bootstrap" Components Nor A "ngDoBootstrap" Method Please define one of these.
+
+@description
+This error mean the Module declared to be bootstrapped in the `main.ts` file (usually the `AppModule`) has empty `bootstrap` array or the `bootstrap` property is missing altogether
```
```suggestion
This error means that an NgModule that was used for bootstrapping an application is missing key information for Angular to proceed with the bootstrap process.

The error usually happens when the NgModule is missing the `bootstrap` property in the `@NgModule` annotation or there is no `ngDoBootstrap` lifecycle hook defined on that NgModule class. More information about the bootstrapping process can be found in [this guide](guide/bootstrapping).
```

robertIsaac replied:
saying `or` means both should be presented, so I made it `and`

also added the possibility that the  `bootstrap` property having empty array is same as being missing, had to alter your sentence a bit to add it this info

AndrewKushnir commented on aio/content/errors/NG0403.md:
```diff
@@ -0,0 +1,30 @@
+@description
+This error mean the Module declared to be bootstrapped in the `main.ts` file (usually the `AppModule`) has empty `bootstrap` array or the `bootstrap` property is missing altogether
+
+both of these examples will throw the error
```
```suggestion
The following examples will trigged the error.
```

robertIsaac replied:
done

AndrewKushnir commented on aio/content/errors/NG0403.md:
````diff
@@ -0,0 +1,30 @@
+both of these examples will throw the error
+
+```typescript
+@NgModule({
+  declarations: [AppComponent],
+  imports: [BrowserModule, AppRoutingModule],
+  providers: [],
+})
+export class AppModule {}
````
```suggestion
@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule, AppRoutingModule],
  providers: [],
})
export class AppModule {}

// The `AppModule` is used for bootstrapping, but the `@NgModule.bootstrap` field is missing.
platformBrowser().bootstrapModule(AppModule);
```

robertIsaac replied:
done

AndrewKushnir commented on aio/content/errors/NG0403.md:
````diff
@@ -0,0 +1,30 @@
+```
+
+```typescript
+@NgModule({
+  declarations: [AppComponent],
+  imports: [BrowserModule, AppRoutingModule],
+  providers: [],
+  bootstrap: [],
+})
+export class AppModule {}
````
```suggestion
@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule, AppRoutingModule],
  providers: [],
  bootstrap: [],
})
export class AppModule {}

// The `AppModule` is used for bootstrapping, but the `@NgModule.bootstrap` field contains an empty array.
platformBrowser().bootstrapModule(AppModule);
```

robertIsaac replied:
done

AndrewKushnir commented on aio/content/errors/NG0403.md:
````diff
@@ -0,0 +1,30 @@
+export class AppModule {}
+```
+
+@debugging
+Did you mistakenly removed the `bootstrap` property of your `AppModule` or remove the `AppComponent` from it?
````
```suggestion
@debugging
Please make sure that the NgModule that is used for bootstrapping is setup correctly:
- either the `bootstrap` property exists (and contains a non-empty array) in the `@NgModule` annotation
- or the `ngDoBootstrap` method exists on the NgModule class

You can find additional information on the bootstrapping process in [this guide](guide/bootstrapping#the-bootstrap-array). 

```

robertIsaac replied:
> You can find additional information on the bootstrapping process in [this guide](guide/bootstrapping#the-bootstrap-array).

is duplicate of same info given in the description, so I didn't add this part

> 

my pleasure to contribute
this change is included now

AndrewKushnir reviewed (commented):
@robertIsaac thanks for creating a PR, I've left some comments.

Could you please also update the `packages/core/src/errors.ts` file to annotate the error code (with the `-` sign), so that the runtime can add a link to that guide when an error is thrown. Here is a diff:

```
--- a/packages/core/src/errors.ts
+++ b/packages/core/src/errors.ts
@@ -53,7 +53,7 @@ export const enum RuntimeErrorCode {
   MULTIPLE_PLATFORMS = 400,
   PLATFORM_NOT_FOUND = 401,
   ERROR_HANDLER_NOT_FOUND = 402,
-  BOOTSTRAP_COMPONENTS_NOT_FOUND = 403,
+  BOOTSTRAP_COMPONENTS_NOT_FOUND = -403,
   PLATFORM_ALREADY_DESTROYED = 404,
   ASYNC_INITIALIZERS_STILL_RUNNING = 405,
   APPLICATION_REF_ALREADY_DESTROYED = 406,
```

Thank you.

robertIsaac commented:
> @robertIsaac thanks for creating a PR, I've left some comments.
> 
> Could you please also update the `packages/core/src/errors.ts` file to annotate the error code (with the `-` sign), so that the runtime can add a link to that guide when an error is thrown. Here is a diff:
> 
> ```
> --- a/packages/core/src/errors.ts
> +++ b/packages/core/src/errors.ts
> @@ -53,7 +53,7 @@ export const enum RuntimeErrorCode {
>    MULTIPLE_PLATFORMS = 400,
>    PLATFORM_NOT_FOUND = 401,
>    ERROR_HANDLER_NOT_FOUND = 402,
> -  BOOTSTRAP_COMPONENTS_NOT_FOUND = 403,
> +  BOOTSTRAP_COMPONENTS_NOT_FOUND = -403,
>    PLATFORM_ALREADY_DESTROYED = 404,
>    ASYNC_INITIALIZERS_STILL_RUNNING = 405,
>    APPLICATION_REF_ALREADY_DESTROYED = 406,
> ```
> 
> Thank you.

my pleasure to contribute
I have addressed all of your comments and made the change to `packages/core/src/errors.ts` you mentioned

AndrewKushnir commented:
@robertIsaac thanks for addressing the feedback! There are a couple small problems identified by the CI:

- It looks like one of the tests should be updated (see https://github.com/angular/angular/blob/main/packages/core/test/application_ref_spec.ts#L451). The test uses the entire error message in the assertion, so it's now failing due to an extra link to the guide. Could you please update that test to include the link?

Here is an error message from the CI:
```
Expected 'NG0403: The module MyModule was bootstrapped, but it does not declare "@NgModule.bootstrap" 
components nor a "ngDoBootstrap" method. Please define one of these. Find more at 
https://angular.io/errors/NG0403' to equal 'NG0403: The module MyModule was bootstrapped, but it does not declare 
"@NgModule.bootstrap" components nor a "ngDoBootstrap" method. Please define one of these.'.
```

- You'd also need to update a public API golden file by running this command: `yarn public-api:update`. As a result, there will be a change within the `goldens` folder, please include that change into the PR as well.

Thank you.

robertIsaac commented:
> @robertIsaac thanks for addressing the feedback! There are a couple small problems identified by the CI:
> 
> - It looks like one of the tests should be updated (see https://github.com/angular/angular/blob/main/packages/core/test/application_ref_spec.ts#L451). The test uses the entire error message in the assertion, so it's now failing due to an extra link to the guide. Could you please update that test to include the link?
> 
> Here is an error message from the CI:
> ```
> Expected 'NG0403: The module MyModule was bootstrapped, but it does not declare "@NgModule.bootstrap" 
> components nor a "ngDoBootstrap" method. Please define one of these. Find more at 
> https://angular.io/errors/NG0403' to equal 'NG0403: The module MyModule was bootstrapped, but it does not declare 
> "@NgModule.bootstrap" components nor a "ngDoBootstrap" method. Please define one of these.'.
> ```
> 
> - You'd also need to update a public API golden file by running this command: `yarn public-api:update`. As a result, there will be a change within the `goldens` folder, please include that change into the PR as well.
> 
> Thank you.

I will do them tomorrow morning
Can you give me a link to see the error message, or how to run it locally to be sure it's working correctly

AndrewKushnir commented:
> Can you give me a link to see the error message, or how to run it locally to be sure it's working correctly

You can try running `yarn bazel test packages/core/test` to reproduce the failure locally. The error message can also be found in CI job outpout ([see here](https://app.circleci.com/pipelines/github/angular/angular/54617/workflows/ef5b6038-2d83-4d04-9588-8c499bc2e7e3/jobs/1272129)). Please let us know if that worked and if any additional information is needed. Thank you.

robertIsaac commented:
> 

thank you for the info
I have fixed the 2 problems now

robertIsaac commented:
@AndrewKushnir I see that CI failed on `aio:e2e`
I ran it locally using `yarn bazel test aio:e2e` and it succeeded
so I'm not sure if it's flaky test, or there is a problem
I tried to rerun the test on CircleCI, but all options are grayed for me
can you please rerun it for me, and if it still fail let me know how to fix it

pkozlowski-opensource commented:
@robertIsaac I've restarted the failing job, but yeh, it has been flaky lately....

robertIsaac commented:
> @robertIsaac I've restarted the failing job, but yeh, it has been flaky lately....

thank you
it has passed now

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

dylhunn reviewed (approved):
reviewed-for: fw-core, public-api

AndrewKushnir commented:
@robertIsaac thanks for addressing the feedback and updating the test! The CI is "green" and this PR received the necessary approvals, so I'm adding it to the merge queue.

Thanks for contributing to Angular and making the docs better 👍
